File: c++/hangman/HangmanGame.cpp
#include "HangmanGame.h"
#include "HangmanRenderer.h"

#include <algorithm>
#include <cctype>
#include <iostream>
#include <set>
#include <string>

using namespace std;

HangmanGame::HangmanGame(const string &secret,
                         const HangmanRenderer &renderer)
    : word_(secret), renderer_(renderer), mistakes_(0),
      maxMistakes_(renderer.maxMistakes()) {}

char HangmanGame::readLetter()
{
    while (true)
    {
        cout << "Enter a letter (A-Z): ";
        string s;
        if (!getline(cin, s))
            return '\0'; // EOF -> quit
        // trim spaces
        s.erase(remove_if(s.begin(), s.end(),
                          [](unsigned char c)
                          { return isspace(c); }),
                s.end());
        if (s.size() == 1 && isalpha(static_cast<unsigned char>(s[0])))
            return static_cast<char>(tolower(static_cast<unsigned char>(s[0])));
        cout << "Invalid input. Please type a single letter.\n";
    }
}

bool HangmanGame::allRevealed() const
{
    for (size_t i = 0; i < word_.size(); ++i)
    {
        char c = static_cast<char>(tolower(static_cast<unsigned char>(word_[i])));
        if (!guessed_.count(c))
            return false;
    }
    return true;
}

string HangmanGame::maskedWord() const
{
    string res;
    for (size_t i = 0; i < word_.size(); ++i)
    {
        char c = word_[i];
        if (guessed_.count(c))
            res.push_back(static_cast<char>(toupper(c)));
        else
            res.push_back('_');
        res.push_back(' ');
    }
    return res;
}

void HangmanGame::printState() const
{
    cout << "\n"
         << renderer_.stage(mistakes_) << "\n";
    cout << "Word: " << maskedWord() << "\n";

    cout << "Guessed: ";
    if (guessed_.empty())
        cout << "(none)";
    else
    {
        for (set<char>::const_iterator it = guessed_.begin();
             it != guessed_.end(); ++it)
        {
            cout << (char)toupper(*it) << ' ';
        }
    }
    cout << "\n";

    cout << "Wrong:   ";
    if (wrong_.empty())
        cout << "(none)";
    else
    {
        for (set<char>::const_iterator it = wrong_.begin();
             it != wrong_.end(); ++it)
        {
            cout << (char)toupper(*it) << ' ';
        }
    }
    cout << "\n";

    cout << "Lives remaining: " << (maxMistakes_ - mistakes_) << "\n";
}

bool HangmanGame::runRound()
{
    cout << "\n========== HANGMAN ==========\n";
    printState();

    while (mistakes_ < maxMistakes_)
    {
        char g = readLetter();
        if (g == '\0')
        {
            cout << "\nGoodbye!\n";
            return false; // quit early
        }

        if (guessed_.count(g) || wrong_.count(g))
        {
            cout << "You already tried '" << (char)toupper(g)
                 << "'. Try another.\n";
            continue;
        }

        if (word_.find(g) != string::npos)
        {
            guessed_.insert(g);
            cout << endl;
            cout << "Nice! '" << (char)toupper(g) << "' is in the word.\n";
        }
        else
        {
            wrong_.insert(g);
            mistakes_++;
            cout << endl;
            cout << "Oops! '" << (char)toupper(g) << "' is not in the word.\n";
        }

        printState();

        if (allRevealed())
        {
            cout << "You WIN! The word was: " << word_ << "\n";
            return true;
        }
    }

    cout << "You LOSE! The word was: " << word_ << "\n";
    return true;
}

File: python/hangman/hangman_game.py
import string

class HangmanGame:
    def __init__(self, secret: str, renderer):
        self.word = secret
        self.renderer = renderer
        self.guessed = set()
        self.wrong = set()
        self.mistakes = 0
        self.max_mistakes = renderer.max_mistakes()

    def read_letter(self):
        while True:
            try:
                s = input("Enter a letter (A-Z): ").strip().lower()
            except EOFError:
                return None
            if len(s) == 1 and s in string.ascii_lowercase:
                return s
            print("Invalid input. Please type a single letter.")

    def all_revealed(self):
        return all(c in self.guessed for c in self.word)

    def masked_word(self):
        return " ".join(c.upper() if c in self.guessed else "_" for c in self.word)

    def print_state(self):
        print("\n" + self.renderer.stage(self.mistakes))
        print(f"Word: {self.masked_word()}")
        print("Guessed:", " ".join(sorted(c.upper() for c in self.guessed)) or "(none)")
        print("Wrong:  ", " ".join(sorted(c.upper() for c in self.wrong)) or "(none)")
        print(f"Lives remaining: {self.max_mistakes - self.mistakes}")

    def run_round(self):
        print("\n========== HANGMAN ==========")
        self.print_state()

        while self.mistakes < self.max_mistakes:
            g = self.read_letter()
            if g is None:
                print("\nGoodbye!")
                return False

            if g in self.guessed or g in self.wrong:
                print(f"You already tried '{g.upper()}'. Try another.")
                continue

            if g in self.word:
                self.guessed.add(g)
                print(f"\nNice! '{g.upper()}' is in the word.\n")
            else:
                self.wrong.add(g)
                self.mistakes += 1
                print(f"\nOops! '{g.upper()}' is not in the word.\n")

            self.print_state()

            if self.all_revealed():
                print(f"You WIN! The word was: {self.word}")
                return True

        print(f"You LOSE! The word was: {self.word}")
        return True
